### PR TITLE
refactor DataSpaceOperations

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- remove the method `map()` from `DataSpaceOperations`
- refcator `linearize()` for `math::Vector` and provide `mapToND()`
- use `linearize()` and `mapToND()` as replacement for `DataSpaceOperations::map()`

By using a better name instead of `map` for mapping to an ND-INdex and for linearizing an ND-Index the code is much better readable. Both functions work with N-dimensional Vectors and not only with up to 3-Dimensions.

- [x] add documentation to `linearize()` and `mapToND`

The register footprint on NVIDIA GPUs stays nearly equal to the original code.

CC-ing: @ikbuibui The upcoming refactoring could be interesting for your plugin.

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
note to myself: The class `MapToPos` should be removed too!

File: include/pmacc/math/vector/Vector.hpp
Comment (ikbuibui) [inline include/pmacc/math/vector/Vector.hpp:761]: T_IntegralType linearIdx{idx[T_dim - 1]};
            for(int i = T_dim - 2; i >= 0; i--)
                linearIdx = linearIdx * pitchExtents[i] + idx[i];
            ```suggestion
            auto pitchExtents = size;
            for(uint32_t d = 1u; d < T_dim - 1u; ++d)
                pitchExtents[d] *= pitchExtents[d - 1u];

            T_IntegralType linearIdx = idx[0];
            for(uint32_t d = 1u; d < T_dim; ++d)
                linearIdx += pitchExtents[d - 1u] * idx[d];
            ```

File: include/pmacc/math/vector/Vector.hpp
Comment (ikbuibui) [inline include/pmacc/math/vector/Vector.hpp:749]: T_IntegralType linearIdx{idx[T_dim - 1]};
        for(int i = T_dim - 2; i >= 0; i--)
            linearIdx = linearIdx * size[i] + idx[i];

Comment (ikbuibui) [reply include/pmacc/math/vector/Vector.hpp:749]: An alternative implementation can be like this. This formulation should also work for mapToND

Comment (psychocoderHPC) [reply include/pmacc/math/vector/Vector.hpp:749]: added your suggestions.

File: include/pmacc/math/vector/Vector.hpp
Comment (PrometheusPi) [inline include/pmacc/math/vector/Vector.hpp]:
```suggestion
        /** Give the linear index of an N-dimensional index within an N-dimensional index space.
```

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
just a very minor typo

**Comment by ikbuibui** [comment]
I think it would be helpful to edit `docs/source/prgpatterns/lockstep.rst` to reflect the change that the linearIdx from the lockstep should be handled as int32_t instead of uint32_t.

File: include/pmacc/math/vector/Vector.hpp
Comment (psychocoderHPC) [reply include/pmacc/math/vector/Vector.hpp]: fixed

**Comment by psychocoderHPC** [comment]
> I think it would be helpful to edit `docs/source/prgpatterns/lockstep.rst` to reflect the change that the linearIdx from the lockstep should be handled as int32_t instead of uint32_t.

Changing the first lambda parameter to int is not required. In the examples, we do not use the linearizing functions.
Using `int` is only required in case you call methods that require `int` and you do not static cast.

**Comment by psychocoderHPC** [comment]
@PrometheusPi PR can be merged

**Comment by ikbuibui** [comment]
The binning plugin is still using the DataSpaceOperations map function (pre refactor), the occurrence needs to be fixed before merging

**Comment by psychocoderHPC** [comment]
> The binning plugin is still using the DataSpaceOperations map function (pre refactor), the occurrence needs to be fixed before merging

Fixed, @PrometheusPi PR can be merged after the CI passed

**Review by PrometheusPi — APPROVED** [review]

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
There is a compile bug in:
```
/builds/hzdr/crp/picongpu/include/picongpu/../picongpu/param/binningSetup.param:55:68: error: no member named 'map' in 'pmacc::DataSpaceOperations<2U>'
                    = pmacc::DataSpaceOperations<SIMDIM>::template map<SuperCellSize>(linearCellIdx);
 ```
 `map` is not known. 
 I think this file was just forgotten.

**Comment by psychocoderHPC** [comment]
> There is a compile bug in:
> 
> ```
> /builds/hzdr/crp/picongpu/include/picongpu/../picongpu/param/binningSetup.param:55:68: error: no member named 'map' in 'pmacc::DataSpaceOperations<2U>'
>                     = pmacc::DataSpaceOperations<SIMDIM>::template map<SuperCellSize>(linearCellIdx);
> ```
> 
> `map` is not known. I think this file was just forgotten.

should be fixed now

**Review by PrometheusPi — APPROVED** [review]

**Pull request merged**
